### PR TITLE
feat: Parse EXT-X-DATERANGE tags in hls_parser

### PR DESCRIPTION
**Pull request by mleung-harmonicinc**
For HMS-7496 HLS graphics overlay.
SOI: https://confluence360.harmonicinc.com/display/~mleung/%28SOI%29+As+a+PPO%2C+I+would+like+to+demo+Channels+Assembly+GFx+signaling+in+HLS+with+HPS-Web

### CODE REVIEW

File: lib/hls/hls_parser.js [hunk 1]
```diff
@@ -1695,6 +1695,22 @@ shaka.hls.HlsParser = class {
...
+    const dateRangeTags = [];
+    if (playlist.segments) {
+      for (const segment of playlist.segments) {
+        const segmentDateRangeTags = shaka.hls.Utils.filterTagsByName(
+            segment.tags, 'EXT-X-DATERANGE');
+        dateRangeTags.push(...segmentDateRangeTags);
+      }
+    }
+
+    const syncTime = segmentIndex.earliestReference().syncTime;
```
Comment (acheung-harmonicinc) [inline lib/hls/hls_parser.js]: `syncTime` here is the first segment's PDT in epoch seconds?

why do we need to subtract it here, but in dash case the startTime returning doesn't need to subtract any offset? 
https://github.com/harmonicinc-com/shaka-player/blob/204e2a0fbb7aef7964e03d4dc8ad3652e826f20d/lib/dash/dash_parser.js#L1976

does syncTime increase as older segments are removed from playlist?

Comment (mleung-harmonicinc) [reply lib/hls/hls_parser.js]: In the HLS date range tag, the `START-DATE` is a datetime, which is different from the time values used in the `TimelineRegionInfo` (the startTime in the beginning is set to 0). In order for the observer to correctly be notified of events such as entering/exiting a region, the datetime has to be offset to reflect the video time.

Yes, the syncTime does increase when the playlist is updated. The previous implementation is incorrect. Updating it now to account for when segments are removed from the playlist.

File: lib/hls/hls_parser.js [hunk 2]
```diff
@@ -2305,6 +2311,46 @@ shaka.hls.HlsParser = class {
...
+                    .plannedDuration;
+              }
+            }
+
+            // Update this region because the endTime (duration) has changed
+            this.playerInterface_.onTimelineRegionAdded(existingRegion);
+          }
+        } else {
+          this.playerInterface_.onTimelineRegionAdded(newRegion);
+          this.regionIdToRegionInfosMap_.set(newRegion.id, newRegion);
```
Comment (acheung-harmonicinc) [inline lib/hls/hls_parser.js:2360]: Do we need to remove old entries from regionIdToRegionInfosMap_ after they moved out of DVR window?

Comment (mleung-harmonicinc) [reply lib/hls/hls_parser.js:2360]: Yes, they should be removed. Added the checking for removal.

**Pull request merged**